File: packages/javascript/src/resources/agents.ts
/**
 * Agent resource — CRUD plus library and workforce-membership operations.
 */

import { del, get, getList, post, put } from "../http.js";
import type {
  Agent,
  AgentCreate,
  AgentUpdate,
  ResolvedConfig,
  WorkforceMember,
} from "../types.js";

export class AgentResource {
  private readonly config: ResolvedConfig;

  constructor(config: ResolvedConfig) {
    this.config = config;
  }

  // ── Core CRUD ────────────────────────────────────────────

  /**
   * List agents, optionally filtered by workforce.
   *
   * @param workforceId - If provided, only agents in this workforce are returned.
   */
  list(workforceId?: string): Promise<Agent[]> {
    const params = workforceId ? { workforce_id: workforceId } : undefined;
    return getList<Agent>(this.config, "/agents", params);
  }

  /** List featured / template agents. */
  listFeatured(): Promise<Agent[]> {
    return getList<Agent>(this.config, "/agents/featured");
  }

  /** Get a single agent by ID. */
  get(agentId: string): Promise<Agent> {
    return get<Agent>(this.config, `/agents/${agentId}`);
  }

  /** Create a new agent. */
  create(data: AgentCreate): Promise<Agent> {
    return post<Agent>(this.config, "/agents", data);
  }

  /** Update an existing agent (server uses PUT). */
  update(agentId: string, data: AgentUpdate): Promise<Agent> {
    return put<Agent>(this.config, `/agents/${agentId}`, data);
  }

  /** Delete an agent. */
  delete(agentId: string): Promise<void> {
    return del<void>(this.config, `/agents/${agentId}`);
  }

  // ── Library operations ──────────────────────────────────

  /** Clone a featured/template agent into the user's library. */
  clone(agentId: string, name?: string): Promise<Agent> {
    const body: Record<string, unknown> = {};
    if (name !== undefined) {
      body["name"] = name;
    }
    return post<Agent>(this.config, `/agents/${agentId}/clone`, body);
  }

  /** Export an agent (returns the raw export envelope). */
  export(agentId: string): Promise<Record<string, unknown>> {
    return get<Record<string, unknown>>(
      this.config,
      `/agents/${agentId}/export`,
    );
  }

  // ── Workforce membership ────────────────────────────────

  /** List agents attached to a workforce, with per-workforce metadata. */
  listWorkforceMembers(workforceId: string): Promise<WorkforceMember[]> {
    return get<WorkforceMember[]>(
      this.config,
      `/workforces/${workforceId}/members`,
    );
  }

  /** Attach a library agent to a workforce. */
  attachToWorkforce(
    workforceId: string,
    agentId: string,
    options: {
      override_json?: Record<string, unknown>;
      position_x?: number;
      position_y?: number;
    } = {},
  ): Promise<WorkforceMember> {
    return post<WorkforceMember>(
      this.config,
      `/workforces/${workforceId}/members`,
      {
        agent_id: agentId,
        override_json: options.override_json ?? {},
        position_x: options.position_x ?? 0,
        position_y: options.position_y ?? 0,
      },
    );
  }

  /** Detach an agent from a workforce. */
  detachFromWorkforce(workforceId: string, agentId: string): Promise<void> {
    return del<void>(
      this.config,
      `/workforces/${workforceId}/members/${agentId}`,
    );
  }
}

File: packages/python/navaia_forge/resources/agents.py
"""Agent resource."""

from __future__ import annotations

from typing import Any

from ..types import Agent, ModelProvider
from ._base import ResourceBase, parse_list, parse_model


class AgentsResource(ResourceBase):
    """CRUD operations for agents."""

    def list(self, workforce_id: str | None = None) -> list[Agent]:
        """List agents, optionally scoped to a workforce."""
        params: dict[str, Any] = {}
        if workforce_id is not None:
            params["workforce_id"] = workforce_id
        return parse_list(Agent, self._http.get_list("/agents", params=params or None))

    def get(self, agent_id: str) -> Agent:
        """Fetch a single agent by id."""
        return parse_model(Agent, self._http.get(f"/agents/{agent_id}"))

    def create(
        self,
        workforce_id: str,
        name: str,
        role: str,
        instructions: str,
        *,
        model_provider: ModelProvider = "anthropic",
        model_name: str = "sonnet",
        max_turns: int = 25,
        tools: list[dict[str, Any]] | None = None,
        knowledge_bases: list[str] | None = None,
        position_x: float = 0,
        position_y: float = 0,
        config_json: dict[str, Any] | None = None,
    ) -> Agent:
        """Create an agent inside a workforce."""
        body: dict[str, Any] = {
            "workforce_id": workforce_id,
            "name": name,
            "role": role,
            "instructions": instructions,
            "model_provider": model_provider,
            "model_name": model_name,
            "max_turns": max_turns,
            "tools": list(tools) if tools is not None else [],
            "knowledge_bases": list(knowledge_bases) if knowledge_bases is not None else [],
            "position_x": position_x,
            "position_y": position_y,
        }
        if config_json is not None:
            body["config_json"] = config_json
        return parse_model(Agent, self._http.post("/agents", body))

    def update(self, agent_id: str, **fields: Any) -> Agent:
        """Patch the agent with the supplied fields."""
        return parse_model(Agent, self._http.patch(f"/agents/{agent_id}", fields))

    def delete(self, agent_id: str) -> None:
        """Delete an agent."""
        self._http.delete(f"/agents/{agent_id}")
